include/xtensor-zarr/xzarr_aws_store.hpp

```diff
@@ -0,0 +1,286 @@
+        {
+            auto err = outcome.GetError();
+            XTENSOR_THROW(std::runtime_error, std::string("Error: GetObject: ") + err.GetExceptionName().c_str() + ": " + err.GetMessage().c_str());
+        }
+
+        auto& reader = outcome.GetResultWithOwnership().GetBody();
+        std::string bytes(std::istreambuf_iterator<char>(reader), {});
+
+        return bytes;
+    }
+
+    void xzarr_aws_stream::operator=(const std::vector<char>& value)
+    {
+        assign(value.data(), value.size());
+    }
+
+    void xzarr_aws_stream::operator=(const std::string& value)
+    {
+        assign(value.c_str(), value.size());
+    }
+
+    void xzarr_aws_stream::assign(const char* value, std::size_t size)
+    {
+        Aws::S3::Model::PutObjectRequest request;
+        request.SetBucket(m_bucket);
+        request.SetKey(m_path);
+
+        std::shared_ptr<Aws::IOStream> writer = Aws::MakeShared<Aws::FStream>("SampleAllocationTag", m_path.c_str(), std::ios_base::in | std::ios_base::binary);
+        writer->write(value, size);
+        writer->flush();
+
+        request.SetBody(writer);
+
+        Aws::S3::Model::PutObjectOutcome outcome = m_client.PutObject(request);
+
+        if (!outcome.IsSuccess())
+        {
+            auto err = outcome.GetError();
+            XTENSOR_THROW(std::runtime_error, std::string("Error: PutObject: ") + err.GetExceptionName().c_str() + ": " + err.GetMessage().c_str());
+        }
+    }
+
+    /**********************************
+     * xzarr_aws_store implementation *
+     **********************************/
+
+    xzarr_aws_store::xzarr_aws_store(const std::string& root, Aws::S3::S3Client& client)
+        : m_root(root)
+        , m_client(client)
+    {
+        if (m_root.empty())
+        {
+            XTENSOR_THROW(std::runtime_error, "Root directory cannot be empty");
+        }
+        while (m_root.back() == '/')
+        {
+            m_root.pop_back();
+        }
+        std::size_t i = m_root.find('/');
+        if (i == std::string::npos)
+        {
+            m_bucket = m_root.c_str();
+            m_root = "";
+        }
+        else
+        {
+            m_bucket = m_root.substr(0, i).c_str();
+            m_root = m_root.substr(i + 1);
+        }
+    }
+
+    xzarr_aws_stream xzarr_aws_store::operator[](const std::string& key)
+    {
+        return xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client);
+    }
+
+    void xzarr_aws_store::set(const std::string& key, const std::vector<char>& value)
+    {
+        xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client) = value;
+    }
+
+    void xzarr_aws_store::set(const std::string& key, const std::string& value)
+    {
+        xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client) = value;
+    }
+
+    std::string xzarr_aws_store::get(const std::string& key)
+    {
+        return std::move(xzarr_aws_stream((m_root + '/' + key).c_str(), m_bucket, m_client));
```

Comment:
Avoid returning `std::move`, return the temporary. The compiler will perform Return Value Optimization and will avoid the copy (it will directly write the result where it has to).